Userland/Libraries/LibShell/Builtin.cpp

```diff
@@ -460,20 +461,37 @@ ErrorOr<int> Shell::builtin_cd(Main::Arguments arguments)
     if (cd_history.is_empty() || cd_history.last() != real_path)
         cd_history.enqueue(real_path);
 
-    auto path_relative_to_current_directory = LexicalPath::relative_path(real_path, cwd);
-    if (path_relative_to_current_directory.is_empty())
-        path_relative_to_current_directory = real_path;
-    char const* path = path_relative_to_current_directory.characters();
+    ByteString path_to_chdir;
+    if (!requested_path_is_absolute)
+        path_to_chdir = LexicalPath::relative_path(real_path, cwd);
+    else
+        path_to_chdir = real_path;
+    char const* path = path_to_chdir.characters();
 
-    int rc = chdir(path);
-    if (rc < 0) {
-        if (errno == ENOTDIR) {
-            warnln("Not a directory: {}", path);
-        } else {
-            warnln("chdir({}) failed: {}", path, strerror(errno));
+    Function<int(bool)> do_chdir = [&](bool tried_lexical_resolution) -> int {
+        int rc = chdir(path);
+        if (rc < 0) {
+            if (errno == ENOTDIR) {
+                warnln("Not a directory: {}", path);
+            } else {
+                if (!tried_lexical_resolution) {
+                    // Resolve lexically then try again.
+                    auto new_path = LexicalPath::join(cwd, path_to_chdir).string();
```

Comment:
This still tries to chdir to the wrong directory in this situation:
```
s@spc:/home/s/repos/serenity/Build/lagom$> cd /root
chdir(/home/s/repos/serenity/Build/lagom/root) failed: No such file or directory
```
I think you should only do this if `path_to_chdir` is relative.